src/edit.rs

```diff
@@ -0,0 +1,159 @@
+
+    debug!("created tmpfile at: {}", tmpf.path().to_str().unwrap());
+    editor.1.push(tmpf.path().to_str().unwrap().to_string());
+    let mut cmd = Command::new(editor.0);
+    cmd.args(editor.1);
+
+    let mut child = cmd
+        .spawn()
+        .map_err(|e| format!("error spawning editor: {}", e))?;
+
+    let exit = child
+        .wait()
+        .map_err(|e| format!("error waiting for editor: {}", e))?;
+
+    if !exit.success() {
+        return Err(format!("editor exited non-zero: {}", exit.code().unwrap()))?;
+    }
+    tmpf.seek(SeekFrom::Start(0))
+        .map_err(|e| format!("could not seek in tempfile: {}", e))?;
+    let mut new_contents = String::new();
+    tmpf.read_to_string(&mut new_contents)
+        .map_err(|e| format!("error reading file: {}", e))?;
+
+    if new_contents.trim() == serialized_data.trim() {
+        debug!("identical data read; shortcutting out");
+        return Ok(PathFrecencyDiff::new(Vec::new(), Vec::new()));
+    }
+
+    let mut new_map = deserialize(&new_contents)?;
+
+    let mut removals = Vec::new();
+    let mut additions = Vec::new();
+
+    // Find all items in the input set and see if they've changed
+    for item in data {
+        debug!("edit: processing {:?}", item);
+        match new_map.remove(&item.0) {
+            Some(w) => {
+                if item.1 != w {
+                    debug!("edit: update {:?}", item.0);
+                    // weight edited, aka remove + add
+                    removals.push(item.0.clone());
+                    additions.push((item.0.clone(), w));
+                }
+                // otherwise, no diff
+            }
+            None => {
+                debug!("edit: removal {:?}", item.0);
+                removals.push(item.0.clone());
+            }
+        }
+    }
+    // anything that was in the input has been removed from our new_map now, so anything
+    // remaining is an addition to the database
+    for (k, v) in new_map {
+        debug!("edit: add {:?}", k);
+        additions.push((k, v));
+    }
+
+    Ok(PathFrecencyDiff::new(additions, removals))
+}
+
+pub fn serialize(matches: &Vec<(String, f64)>) -> String {
+    format!(
+        r#"# Edit your frecency fearlessly!
+#
+# Lines starting with '#' are comments. sh-esque quoting and escapes may be used in paths.
+# Columns are whitespace separated. The first column is the current score, the second is the path.
+# Any changes saved here will be applied back to your frecency database immediately.
+{}"#,
+        matches
+            .iter()
+            .map(|(s, w)| format!("{}\t{}", w, snailquote::escape(s)))
+            .collect::<Vec<String>>()
+            .join("\n")
+    )
+}
+
+pub fn deserialize(s: &str) -> Result<HashMap<String, f64>, String> {
+    let mut res = HashMap::new();
+    for mut line in s.lines() {
+        if line.len() == 0 || line.starts_with('#') {
+            continue;
+        }
+
+        line = line.trim_right();
```

Comment:
1) Why not just `trim()`?
2) If you move this above line 135, it handles the case of lines with just whitespace instead of erroring on them.